cmd/alert-healer/main.go

```diff
@@ -92,18 +91,25 @@ func main() {
 		}
 	}
 
-	// Create the client:
-	client, err := openshift.NewForConfig(config)
+	// Create the Kuberntes API client:
+	k8sClient, err := kubernetes.NewForConfig(config)
 	if err != nil {
-		glog.Fatalf("Error building client: %s", err.Error())
+		glog.Fatalf("Error building Kubernets API client: %s", err.Error())
+	}
+
+	// Create the OpenShift API client:
+	osClient, err := openshift.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("Error building OpenShift API client: %s", err.Error())
 	}
 
 	// Create an informer factory that will create informes that sync every 5 minutes:
-	informerFactory := informers.NewSharedInformerFactory(client, 5*time.Minute)
+	informerFactory := informers.NewSharedInformerFactory(osClient, 5*time.Minute)
 
 	// Build the healer:
 	healer, err := NewHealerBuilder().
-		Client(client).
+		KubernetesClient(k8sClient).
```

Comment:
From what I see in the code, a Healer is always used with `osClient` and `k8sClient` - it is never used as an empty object, if this is the case, these methods seem a bit redundant and can be replaced with a composite literal i.e. `Healer{osClient, k8sClient}`.

Correct me if I am mistaken - `Healer` cannot work without these clients?

Reply:
Yes, the builder and these methods aren't really needed in this case. However, I like that "bulider" idiom because it decouples the caller from the internal implementation details. Also it allows you to build multiple healers with the same builder, it allows you to do complex checks in the `Build` method before actually creating the object, etc.

For example, to address your concern about the fact that all the parameters are needed we could add someting like this inside the `Build` method:

```go
// Check that all the required parameters have been provided:
if b.k8sClient == nil {
  err = fmt.Errorf("The 'KubernetesClient' parameter is mandatory")
  return
}
if b.osClient == nil {
  err = fmt.Errorf("The 'OpenShift' parameter is mandatory")
  return
}
if (b.informerFactory == nil {
  err = fmt.Errorf("The 'InformerFactory' parameter is mandatory")
  return
}
```

We could even use an slice of errors, and generate a better error message that indicates all the missing parameters, not just the first one. This is the kind of flexibility that the builder idiom brings. I'd prefer to keep it, unless you object strongly.

Reply:
> I like that "bulider" idiom because it decouples the caller from the internal implementation details. Also it allows you to build multiple healers with the same builder, it allows you to do complex checks in the Build method before actually creating the object, etc.

Sounds good, I just wasn't aware of those use cases which make sense now :+1: